src/world/partner/kooper.c

```diff
@@ -93,17 +151,543 @@ s32 func_802BD260_31B280(Evt* evt, s32 arg1) {
+                    npc->pos.x = sp20;
+                    npc->pos.y = sp24;
+                    npc->pos.z = sp28;
+                    npc->planarFlyDist += npc->moveSpeed;
+                    func_8003D660(npc, 1);
+                    npc->moveSpeed += 2.0;
+
+                    if (npc->moveSpeed > 14.0) {
+                        npc->moveSpeed = 14.0f;
+                    }
+
+                    if ((func_800397E8(npc, 6.0f) == 0) && ((playerStatus->flags & 6) == 0)) {
+                        npc->pos.y = (npc->pos.y + ((playerStatus->position.y - npc->pos.y) / 10.0f));
+                    }
+
+                    npc_do_other_npc_collision(npc);
+
+                    if ((npc->flags & 0x2000000) == 0) {
+                        if (func_802BD17C_31B19C(npc) != 0) {
+                            sfx_play_sound_at_npc(0x286, 0, -4);
+                            temp_f20_4 = sin_deg(npc->yaw);
+                            playFX_18(3, npc->pos.x, npc->pos.y + npc->collisionHeight, npc->pos.z, temp_f20_4, -1.0f, -cos_deg(npc->yaw), 1);
+                            sfx_play_sound_at_npc(0, 0, -4);
+                            script->functionTemp[0] = 4;
+                            D_802BEC50 = 8;
+                            npc->moveSpeed -= 4.0;
+                            if (npc->moveSpeed < 0.01) {
+                                npc->moveSpeed = 0.01f;
+                                npc->planarFlyDist += 1.0;
+                            }
+                        }
+                        else if (D_802BEB40_31CB60 == 2) {
+                            sfx_play_sound_at_npc(0, 0, -4);
+                            script->functionTemp[0] = 5;
+                            D_802BEC50 = 30;
+                            npc->moveSpeed = 0.0f;
+                        } else {
+                            if (npc->planarFlyDist > 140.0f) {
+                                label2:
+                                script->functionTemp[0] = 7;
+                                npc->moveSpeed = 0.0f;
+                                sfx_play_sound_at_npc(0, 0, -4);
+                            }
+                            
+                            else if (npc->planarFlyDist > 105.0f) {
+                                npc->moveSpeed -= 4.0;
+                                if (npc->moveSpeed < 0.01) {
+                                    npc->moveSpeed = 0.01f;
+                                    npc->planarFlyDist += 1.0;
+                                }
+                            }
+                            break;
+                        }
+                    } else {
+                        goto label2;
```

Comment:
We should really try to get rid of this. have you tried copying the contents of label2 in here?